Codesandbox doesn't work with Bootstrap SCSS imports

# 🐛 bug report

## Preflight Checklist
<!-- Please ensure you've completed the following steps by replacing [ ] with [x]-->

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

## Description of the problem

CodeSandbox displays the following error when trying to hot load changes to a .scss file that imports Bootstrap: `Cannot find module 'bootstrap/scss/mixins' from '/src'`. The initial transpilation works fine, so it seems to just be an issue with how hot loading handles imports.

I believe the issue is that Bootstrap has both a file named `_mixins.scss` and a directory named `mixins`. The initial transpilation properly imports `_mixins.scss` with the syntax `@import "mixins";`. My best guess is that subsequent hot loading either attempts to import the `mixins` directory instead or it's having issues with the underscore prefix in `_mixins.scss` and it's not finding it.

This affects Bootstrap and presumably other CSS frameworks that have this file/folder structure.

### To Reproduce
1. Visit this Codesandbox: https://codesandbox.io/s/codesandbox-scss-bug-37op3?file=/src/styles.scss
2. Wait for the initial transpilation then modify the primary color SCSS variable to any other color.
3. You'll see the error `Cannot find module 'bootstrap/scss/mixins' from '/src'`
4. Refresh and you'll notice your change takes affect (initial transpilation works, just an issue with hot loading)

### Further Thoughts

I noticed there was previously a fix (https://github.com/codesandbox/codesandbox-client/issues/529) for SASS resolving underscores incorrectly. Maybe this fix isn't present in the hot loading logic?

### Link to sandbox: [link](https://codesandbox.io/s/codesandbox-scss-bug-37op3?file=/src/styles.scss)

### Your Environment

| Software         | Name/Version |
| ---------------- | ------------ |
| Сodesandbox      | 40d75182a
| Browser          | Chrome 84.0.4147.125
| Operating System | macOS 10.15.6

--- a/packages/app/src/sandbox/eval/transpilers/sass/worker/sass-worker.js
+++ b/packages/app/src/sandbox/eval/transpilers/sass/worker/sass-worker.js
@@ -18,18 +18,12 @@ declare var Sass: {
   registerPlugin: (name: string, plugin: Function) => void,
 };
 
-const resolveAsyncModule = (
+function resolveAsyncModule(
   path: string,
-  { ignoredExtensions }?: { ignoredExtensions?: Array<string> }
-) =>
-  new Promise((r, reject) => {
+  { ignoredExtensions }: { ignoredExtensions?: Array<string> } = {}
+) {
+  return new Promise((r, reject) => {
     const sendId = Math.floor(Math.random() * 10000);
-    self.postMessage({
-      type: 'resolve-async-transpiled-module',
-      path,
-      id: sendId,
-      options: { isAbsolute: true, ignoredExtensions },
-    });
 
     const resolveFunc = message => {
       const { type, id, found } = message.data;
@@ -48,12 +42,20 @@ const resolveAsyncModule = (
     };
 
     self.addEventListener('message', resolveFunc);
+
+    self.postMessage({
+      type: 'resolve-async-transpiled-module',
+      path,
+      id: sendId,
+      options: { isAbsolute: true, ignoredExtensions },
+    });
   });
+}
 
 const SUPPORTED_EXTS = ['scss', 'sass', 'css'];
 
-const existsPromise = (fs, file) =>
-  new Promise(r => {
+function existsPromise(fs, file) {
+  return new Promise(r => {
     fs.stat(file, async (err, stats) => {
       if (err || stats.isDirectory()) {
         if (stats && stats.isDirectory()) {
@@ -82,6 +84,7 @@ const existsPromise = (fs, file) =>
       }
     });
   });
+}
 
 /**
  * Return and stop as soon as one promise returns a truthy value
@@ -116,25 +119,23 @@ const getExistingPath = async (fs, p) => {
 };
 
 const resolvedCache = {};
-const resolveSass = (fs, p, path) => {
+async function resolveSass(fs, p, path) {
   const usedPath = p.startsWith('~') ? p.replace('~', '/node_modules/') : p;
 
   const sourceDir = dirname(path);
   resolvedCache[sourceDir] = resolvedCache[sourceDir] || {};
   if (resolvedCache[sourceDir][usedPath]) {
-    return Promise.resolve(resolvedCache[sourceDir][usedPath]);
+    return resolvedCache[sourceDir][usedPath];
   }
 
-  return new Promise((r, reject) => {
-    const directPath = join(sourceDir, usedPath);
+  const directPath = usedPath[0] === '/' ? usedPath : join(sourceDir, usedPath);
 
-    // First try to do the relative path, as a performance optimization
-    getExistingPath(fs, directPath).then(foundPath => {
-      if (foundPath) {
-        r(foundPath);
-        return;
-      }
-
+  // First try to do the relative path, as a performance optimization
+  const existingPath = await getExistingPath(fs, directPath);
+  if (existingPath) {
+    resolvedCache[sourceDir][usedPath] = existingPath;
+  } else {
+    await new Promise((promiseResolve, promiseReject) => {
       resolve(
         usedPath,
         {
@@ -165,23 +166,21 @@ const resolveSass = (fs, p, path) => {
         async (err, resolvedPath) => {
           if (err) {
             if (/^\w/.test(p)) {
-              r(resolveSass(fs, '.' + absolute(p), path));
+              promiseResolve(resolveSass(fs, '.' + absolute(p), path));
+            } else {
+              promiseReject(err);
             }
-
-            reject(err);
           } else {
             const newFoundPath = await getExistingPath(fs, resolvedPath);
-
-            r(newFoundPath);
+            promiseResolve(newFoundPath);
           }
         }
       );
     });
-  }).then(result => {
-    resolvedCache[sourceDir][usedPath] = result;
-    return result;
-  });
-};
+  }
+
+  return resolvedCache[sourceDir][usedPath];
+}
 
 function initializeBrowserFS() {
   return new Promise(res => {
@@ -232,8 +231,8 @@ self.addEventListener('message', async event => {
       const currentPath =
         request.previous === 'stdin' ? path : request.previous;
 
-      const foundPath =
-        request.path || (await resolveSass(fs, request.current, currentPath));
+      // request.path sometimes returns a partially resolved path
+      const foundPath = await resolveSass(fs, request.current, currentPath);
 
       self.postMessage({
         type: 'add-transpilation-dependency',

--- a/packages/app/src/sandbox/eval/transpilers/sass/worker/sass-worker.js
+++ b/packages/app/src/sandbox/eval/transpilers/sass/worker/sass-worker.js
@@ -232,6 +232,7 @@ self.addEventListener('message', async event => {
         request.previous === 'stdin' ? path : request.previous;
 
       // request.path sometimes returns a partially resolved path
+      // See: https://github.com/codesandbox/codesandbox-client/issues/4865
       const foundPath = await resolveSass(fs, request.current, currentPath);
 
       self.postMessage({

